### PR TITLE
[SPARK-37944][SQL] Use error classes in the execution errors of casting

### PR DESCRIPTION
**Pull request by goutam-git**
### What changes were proposed in this pull request?

Migrate the following errors in QueryExecutionErrors:

-  failedToCastValueToDataTypeForPartitionColumnError
- invalidInputSyntaxForNumericError
- cannotCastToDateTimeError
- invalidInputSyntaxForBooleanError
- nullLiteralsCannotBeCastedError

onto use error classes. Throw an implementation of SparkThrowable. Also write a test per every error in QueryExecutionErrorsSuite.
Please Note  invalidInputSyntaxForNumericError and cannotCastToDateTimeError are not present in any more and invalidInputSyntaxForBooleanError is already using error classes.

### Why are the changes needed?

To follow the error class framework and improve test coverage.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

mvn  -Dtest=none -DwildcardSuites=org.apache.spark.sql.errors.QueryExecutionErrorsSuite test

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala [hunk 1]
```diff
@@ -653,26 +653,6 @@ class QueryExecutionErrorsSuite
...
-  test("CAST_OVERFLOW: from long to ANSI intervals") {
-    Seq(
-      LongType -> "9223372036854775807L",
-      DecimalType(19, 0) -> "9223372036854775807BD").foreach { case (sourceType, sourceValue) =>
-      Seq("INTERVAL YEAR TO MONTH", "INTERVAL HOUR TO MINUTE").foreach { it =>
-        checkError(
-          exception = intercept[SparkArithmeticException] {
-            sql(s"select CAST($sourceValue AS $it)").collect()
-          },
-          errorClass = "CAST_OVERFLOW",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:666]: Why do you remove this check for `CAST_OVERFLOW`?

Comment (goutam-git) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala]: It must have got removed during merge. Sorry for the unintentional mistake. Added the check.

**Review by MaxGekk — COMMENTED** [review]
@goutam-git The test failure is related to your changes, probably:
```
FileIndexSuite.SPARK-26263: Throw exception when partition value can't be casted to user-specified type
org.scalatest.exceptions.TestFailedException: "[[FAILED_CAST_PARTITION] Failed to cast value foo to "INT" for partition column a]" did not equal "[Fai
```
Could you fix it, please.

**Comment by goutam-git** [comment]
Not able to build because of 
[error] /home/runner/work/spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:2231:82: no arguments allowed for nullary method elementAtByIndexZeroError: ()org.apache.spark.SparkRuntimeException
[error]             throw QueryExecutionErrors.elementAtByIndexZeroError(getContextOrNull())

@MaxGekk
can you please help resolve it

**Pull request closed without merging**